src/plugins/discord/fetchDiscord.js

```diff
@@ -0,0 +1,198 @@
+import {type endCursor, type pageInfo, Fetcher} from "./fetcher";
+
+const DISCORD_SERVER = "https://discordapp.com/api";
+
+type FetchPlanOptions = {|
+  lastMessageId: string,
+  lastMemberId: string,
+  lastReactionId: string,
+|};
+
+/**
+ * Fetches provided url wwith provided options.
+ * Returns either the response.json or rejects with
+ * a retry flag and an optional message.
+ *
+ */
+function tryDiscordFetch(fetch, url, fetchOptions): Promise<any> {
+  return fetch(url, fetchOptions).then((response) => {
+    switch (response.status) {
+      case 403:
+        return Promise.reject({
+          retry: false,
+          message: `403 Forbidden: bad API username or key?\n${response.url}`,
+        });
+      case 404:
+        return Promise.reject({
+          retry: false,
+          message: `404 Not Found on: ${response.url}; maybe bad serverUrl?`,
+        });
+      case 410:
+        return Promise.reject({retry: false, message: `410 Gone`});
+      case 429:
+        return Promise.reject({retry: true, message: null});
+      default:
+        return Promise.resolve(response.json());
+    }
+  });
+}
+
+// If retry = true in tryDiscordFetch reject, will re-attempt the fetch.
+function retryDiscordFetch(fetch, url, fetchOptions) {
+  return new Promise((resolve, reject) => {
+    const operation = retry.operation();
+    operation.attempt(() => {
+      tryDiscordFetch(fetch, url, fetchOptions)
+        .then((result) => {
+          resolve(result);
+        })
+        .catch((error) => {
+          if (error.retry && operation.retry(true)) {
+            return;
+          } else if (error.message) {
+            reject(new Error(error.message));
+          } else {
+            reject(error);
+          }
+        });
+    });
+  });
+}
+
+/**
+ * Returns a function that ties the fetch options and
+ * Discord server url to the supplied endpoint.
+ * Clients can provide bot token here to create a Discord Fetch.
+ */
+export function makeDiscordFetch(fetch: typeof fetch, token: BotToken) {
+  return async (endpoint: string) => {
+    const fetchOptions = {
+      method: "GET",
+      headers: {
+        Accept: "application/json",
+        Authorization: `Bot ${token}`,
+      },
+    };
+    const url = new URL(`${DISCORD_SERVER}/${endpoint}`).href;
+    return retryDiscordFetch(fetch, url, fetchOptions);
+  };
+}
+
+/**
+ * Iteratively calls provided `fetch` for as long as the `hasNextPage`
+ * variable in the `pageInfo` response is true.
+ *
+ */
+export async function paginatedFetch(
+  fetch: (endCursor: endCursor) => Promise<pageInfo>,
+  endCursor: endCursor
+): Promise<void> {
+  let after = endCursor;
+  let cont = true;
+
+  while (cont) {
+    const {hasNextPage, endCursor} = await fetch(after);
+    after = endCursor;
+    cont = hasNextPage;
+  }
+}
+
+// Verifies that the provided bot credentials have access to the provided Guild
+async function rejectInvalidGuildId(fetcher: Fetcher) {
```

Comment:
This bit adds a more granular error message for clients; the tradeoff in complexity seems minimal. It verifies whether the guild id provided is linked to a guild that actually exists in 3-4 lines. It feels like a plausible error to mis-type or copy/paste a token. 

That said, I don't love this function. I'm don't feel airtight that the logic is correct. I.e, I'm not sure a non-200 response from the Discord server for a guild id means that the Guild doesn't exist.  There also may be additional checks we can perform, perhaps for token-length .

Reply:
Looking again I see what the intention is now.

1. You're looking if the guild is listed in `/users/@me/guilds`.
2. You're checking if the guild even exists with `/guilds/${guildId}`.

The second one could improve the error message if it's not a permission error but it's the wrong ID to begin with.

I've tested the 2nd step with Postman. And it turns out that we always receive a 403 response regardless of whether it exists.

For example:
- `/guilds/12345` is not a valid Snowflake and returns 403.
- `/guilds/9223372036854775807` is the maximum allowed Snowflake and returns 403.
- `/guilds/616186924390023171` does exist (server for [Tauri](https://github.com/tauri-apps/)), but no access returns 403.

(Note, I'm reasonably confident the maximum Snowflake value doesn't exist, given the timestamp it contains when parsing: [see docs](https://discordapp.com/developers/docs/reference#snowflakes) and [parser I used](https://github.com/dirigeants/snowflake/blob/master/src/index.ts))
```js
{
    timestamp: "2084-09-06T15:47:35.551Z",
    workerID: 31,
    processID: 31,
    increment: 4095
}
```

So adding the extra step of querying the specific guild doesn't refine the error.

---

On a different note though, it _is_ a useful URL to query.
Because I just noticed that a `/guild/${guildId}` query already includes roles and custom emoji in the response! Which we are interested in. Definitely the custom emoji.

So we could make this a lot simpler:

```js
// This doubles as a validation we have access.
try {
  const {name, emoji} = await fetch.guild(guildId);
} catch (e) {
  throw new Error(
    `Couldn't fetch Server information, maybe the ID is wrong, ` +
      `or the bot doesn't have access?\n${e.message}`
  );
}
```

Reply:
Fell out of scope.